packages/datadog-plugin-couchbase/src/index.js

```diff
@@ -1,171 +1,91 @@
-      }
-
-      return scope.bind(query.apply(this, arguments))
-    }
+  addSubs (func, start, asyncEnd = defaultAsyncEnd) {
+    this.addSub(`apm:couchbase:${func}:start`, start)
+    this.addSub(`apm:couchbase:${func}:end`, this.exit.bind(this))
+    this.addSub(`apm:couchbase:${func}:error`, errorHandler)
+    this.addSub(`apm:couchbase:${func}:async-end`, asyncEnd)
   }
-}
-
-function createWrapN1qlReq (tracer, config) {
-  return function wrapN1qlReq (_n1qlReq) {
-    return function n1qlReqWithTrace (host, q, adhoc, emitter) {
-      if (!emitter || !emitter.once) return _n1qlReq.apply(this, arguments)
-
-      const scope = tracer.scope()
-      const n1qlQuery = q && q.statement
-      const span = startSpan(tracer, config, 'query', n1qlQuery)
 
-      span.setTag('span.type', 'sql')
-
-      addBucketTag(span, this)
-      onRequestFinish(emitter, span)
-
-      return scope.bind(_n1qlReq, span).apply(this, arguments)
+  startSpan (operation, customTags, store, bucket) {
+    const tags = {
+      'db.type': 'couchbase',
+      'component': 'couchbase',
+      'service.name': this.config.service || `${this.tracer._service}-couchbase`,
+      'resource.name': `couchbase.${operation}`,
+      'span.kind': 'client'
     }
-  }
-}
-
-function createWrapStore (tracer, config, operation) {
-  return function wrapStore (store) {
-    return function storeWithTrace (key, value, options, callback) {
-      const callbackIndex = findCallbackIndex(arguments)
-
-      if (callbackIndex < 0) return store.apply(this, arguments)
-
-      const scope = tracer.scope()
-      const span = startSpan(tracer, config, operation)
-
-      addBucketTag(span, this)
 
-      arguments[callbackIndex] = wrapCallback(span, arguments[callbackIndex])
-
-      return scope.bind(store, span).apply(this, arguments)
+    for (const tag in customTags) {
+      tags[tag] = customTags[tag]
     }
-  }
-}
+    const span = this.tracer.startSpan(`couchbase.${operation}`, {
+      childOf: store ? store.span : null,
+      tags
+    })
 
-function addBucketTag (span, bucket) {
-  span.setTag('couchbase.bucket.name', bucket.name || bucket._name)
-}
+    span.setTag('couchbase.bucket.name', bucket.name || bucket._name)
 
-function findCallbackIndex (args) {
-  for (let i = args.length - 1; i >= 2; i--) {
-    if (typeof args[i] === 'function') return i
+    analyticsSampler.sample(span, this.config.measured)
+    return span
   }
 
-  return -1
+  constructor (...args) {
+    super(...args)
+
+    this.addSubs('_n1qlReq', ([resource, bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('query', { 'span.type': 'sql', 'resource.name': resource }, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('upsert', ([bucket]) => {
+      const store = storage.getStore()
+      const span = this.startSpan('upsert', {}, store, bucket)
+      this.enter(span, store)
+    })
+
+    this.addSubs('insert', ([bucket]) => {
```

Comment:
Looks like a lot of these are not covered by tests. The code looks pretty much always identical, any reason not to have a single handler for most of them?

Reply:
They have different operation names which we use to create a span with, and thus need different channels i'm assuming?

Reply:
Yes but that's the only thing that changes per sub, so it should be possible to reuse the same function and avoid the code coverage issue.